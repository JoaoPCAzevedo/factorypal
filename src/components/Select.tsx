/** Load dependencies */
import React from "react";

/** Interfaces */
interface SelectProps {
  name: string;
  className?: string;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  data: string[];
  selected?: string;
}

/** Component */
const Select: React.FunctionComponent<SelectProps> = ({
  name,
  className,
  data,
  onChange,
  selected,
}) => {
  return (
    <select
      name={name}
      className={className}
      onChange={(e) => onChange(e)}
      data-testid="change-select"
      value={selected}
    >
      {data.map((eachData) => (
        <option key={eachData} value={eachData}>
          {eachData}
        </option>
      ))}
    </select>
  );
};

export default Select;
