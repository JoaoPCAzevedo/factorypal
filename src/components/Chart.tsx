/** Load dependencies */
import React from "react";
import { Chart as ChartJS, ChartType } from "chart.js";
import "chartjs-plugin-datalabels";

/** Load components */
import Select from "./Select";

/** Load types */
import { Metric } from "../types";

/** Types */
type Context =
  | string
  | CanvasRenderingContext2D
  | HTMLCanvasElement
  | ArrayLike<CanvasRenderingContext2D | HTMLCanvasElement>;

type RefContext =
  | string
  | ((instance: HTMLCanvasElement | null) => void)
  | React.RefObject<HTMLCanvasElement>
  | null
  | undefined;

export type ChartColors = {
  backgroundColor: string[];
  borderColor: string[];
};

/** Interfaces */
interface Datalabels {
  anchor: string;
  align: string;
  formatter: (x: number) => number;
  font: {
    weight: string;
  };
}

interface ChartProps {
  className?: string;
  data: {
    labels: Metric["label"][];
    values: Metric["value"][];
    colors?: ChartColors;
  };
  types?: ChartType[];
}

/** Component */
const Chart: React.FunctionComponent<ChartProps> = ({
  className,
  data,
  types,
}) => {
  const [chartType, setChartType] = React.useState<ChartType>("bar");
  const ctx: React.MutableRefObject<Context> = React.useRef<Context>("");

  React.useEffect(() => {
    const chart: Context = ctx.current;

    const currentChart = new ChartJS(chart, {
      plugins: {
        datalabels: {
          anchor: "end",
          align: "top",
          formatter: Math.round,
          font: {
            weight: "bold",
          },
        } as Datalabels,
      } as any,
      type: chartType,
      data: {
        labels: data.labels,
        datasets: [
          {
            data: data.values,
            backgroundColor: data.colors?.backgroundColor,
            borderColor: data.colors?.borderColor,
            borderWidth: 1,
          },
        ],
      },
      options: {
        responsive: true,
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true,
              },
            },
          ],
        },
        legend: {
          display: false,
        },
        animation: {
          duration: 1000,
        },
      },
    });
    return () => {
      currentChart.destroy();
    };
  }, [data, chartType]);

  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setChartType(event.target.value as ChartType);
  };

  return (
    <div className={className}>
      {types && (
        <div className="block is-flex is-justify-content-flex-end is-align-items-center">
          <label className="label mr-4" htmlFor="chart types">
            Chart type:
          </label>
          <div className="select">
            <Select
              name="chart-types"
              className="is-capitalized"
              onChange={(e) => handleChange(e)}
              data={types}
            />
          </div>
        </div>
      )}
      <canvas key="chart" ref={ctx as RefContext} />
    </div>
  );
};

export default Chart;
