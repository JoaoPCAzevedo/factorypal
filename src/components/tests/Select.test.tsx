/** Load dependencies */
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

/** Load components */
import Select from "../Select";

/** Load components */
import { ChartTypes } from "../Chart";

/** Load configs */
const types: ChartTypes[] = [
  "bar",
  "line",
  "radar",
  "pie",
  "doughnut",
  "polarArea",
];

/** Tests */
let container: any = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders with or without a name", () => {
  act(() => {
    render(
      <Select
        name="select-types"
        data={types}
        onChange={(e) => console.log(e.target.value)}
      />,
      container
    );
  });
  expect(container).toMatchSnapshot();
});

it("changes value when clicked", () => {
  const onChange = jest.fn();
  act(() => {
    render(
      <Select name="select-types" data={types} onChange={onChange} />,
      container
    );
  });

  const select: any = document.querySelector("[data-testid=change-select]");

  act(() => {
    select.value = "bar";
  });

  expect(select.value).toBe("bar");
});
