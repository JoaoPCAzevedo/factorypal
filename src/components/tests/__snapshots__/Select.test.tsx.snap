// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with or without a name 1`] = `
<div>
  <select
    data-testid="change-select"
    name="select-types"
  >
    <option
      value="bar"
    >
      bar
    </option>
    <option
      value="line"
    >
      line
    </option>
    <option
      value="radar"
    >
      radar
    </option>
    <option
      value="pie"
    >
      pie
    </option>
    <option
      value="doughnut"
    >
      doughnut
    </option>
    <option
      value="polarArea"
    >
      polarArea
    </option>
  </select>
</div>
`;
